### PR TITLE
.*: generate and embed Openshift's InstallConfig from Tectonic's Config

### PR DESCRIPTION
**Pull request by abhinavdahiya**
embedding the Openshift's InstallConfig in clusters created by tectonic
installer binary allows operators to start building on their config discovery
logic.

### CODE REVIEW

File: installer/pkg/config-generator/fixtures/kube-system.yaml [hunk 1]
```diff
@@ -1,5 +1,49 @@
...
+      serviceCIDR:
+        IP: 10.3.0.0
+        Mask: //8AAA==
+      type: canal
+    platform:
+      aws:
+        region: eu-west-1
+        vpcCIDRBlock: 10.0.0.0/16
+        vpcID: ""
+    pullSecret: /path/config.json
```
Comment (wking) [inline installer/pkg/config-generator/fixtures/kube-system.yaml]: This is no longer useful.  The config should be updated to take the pull secret as a JSON string, even if users are allowed to provide it via a path.

Comment (wking) [reply installer/pkg/config-generator/fixtures/kube-system.yaml]: > just to clarify, the `types.InstallConfig.PullSecret` field is a string type that stores the contents of the pull secret file , which is a JSON string?

That's fine with me.  We expect to consume this as an opaque string, right?  So we should store it as a string (like the SSH pubkey) instead of parsing it out into a more detailed structure.

File: installer/pkg/config-generator/generator_test.go [hunk 2]
```diff
@@ -11,19 +11,31 @@ import (
...
 	if err != nil {
 		t.Errorf("Test case TestUrlFunctions: failed to parse test config, %s", err)
 	}
 
+	tf, err := ioutil.TempFile("", "generator_test")
+	if err != nil {
+		t.Errorf("error creating temp file: %v", err)
+	}
+	if _, err := tf.Write(pullSecretContents); err != nil {
+		t.Errorf("error writing to temp file %q: %v", tf.Name(), err)
```
Comment (wking) [inline installer/pkg/config-generator/generator_test.go]: To clarify [my earlier comments][1], I don't think we want to store the pull secret path at all, and that we should replace `PullSecretPath` with a `PullSecret string` in the `Cluster` type.  Then there would be no need for a temporary file here.  The only filesystem connection would be on input, where we would provide users with the option to load the pull secret from a file.  I set up something similar for SSH pubkeys [here][2].  For pull secrets, I expect the logic to be something like:

```
Pull secret:
Enter your pull secret as a single line of JSON (e.g. '{...}') or as a path to a file containing the pull secret JSON.
```

Then use an "is this valid JSON?" on the result.  If it is, assume it's the pull secret.  If not, assume it's the path and try to load valid JSON from the path.  If you can't, complain about it and post the prompt again.

We have an existing "does this file contain valid JSON?" check [here][3], but I think you'll want something like [this][4] to make it easy to get good invalid-JSON error messages for both strings and files.

[1]: https://github.com/openshift/installer/pull/205/#discussion_r215117166
[2]: https://github.com/openshift/installer/pull/199/files#diff-0fa610c35b30a72dea1ff2a8c0b1e2e9R26
[3]: https://github.com/openshift/installer/blob/f2e6b3ed56aa114cd263369fd626d6b0e493e6ed/installer/pkg/validate/validate.go#L32
[4]: https://github.com/wking/openshift-installer/commit/bd1ebb3ac784c606dcc58aa42cc790513c36306b#diff-3a2d1fe927780885ccdacf5377759554L41

Comment (abhinavdahiya) [reply installer/pkg/config-generator/generator_test.go]: @wking 
I'm sure we can change the `Cluster` type to store the pull secret. But i'm not convinced this the PR to change how we consume the pullsecret in `Cluster`

Comment (wking) [reply installer/pkg/config-generator/generator_test.go]: > I'm sure we can change the `Cluster` type to store the pull secret. But i'm not convinced this the PR to change how we consume the pullsecret in `Cluster`

I've filed #208 with the file -> string change.  I'm fine landing that PR first and then rebasing this one on top (which would let you drop this temp file, etc.).  I'm also fine landing this PR first and rebasing mine on top.  Which way do you want to go?

Comment (abhinavdahiya) [reply installer/pkg/config-generator/generator_test.go]: I can rebase on top of 208

**Comment by abhinavdahiya** [comment]
@wking PTAL updated to use PullSecret string

File: installer/pkg/config-generator/generator.go [hunk 3]
```diff
@@ -95,6 +101,105 @@ func (c *ConfigGenerator) TectonicSystem() (string, error) {
 	})
 }
 
+// InstallConfig returns, if successful, a yaml string for on-disk install-config from 4.0 installer.
```
Comment (wking) [inline installer/pkg/config-generator/generator.go]: I'd reword to be more generic (no need to mention disks or 4.0).  Maybe something like:

> InstallConfig returns a YAML-rendered Kubernetes object with the user-supplied cluster configuration.

File: installer/pkg/config-generator/generator.go [hunk 4]
```diff
@@ -95,6 +101,105 @@ func (c *ConfigGenerator) TectonicSystem() (string, error) {
...
+		Admin: types.Admin{
+			Email:    c.Admin.Email,
+			Password: c.Admin.Password,
+			SSHKey:   c.Admin.SSHKey,
+		},
+		BaseDomain: c.BaseDomain,
+		PullSecret: c.PullSecret,
+	}
+
+	ic.Networking.Type = types.NetworkType(string(c.Networking.Type))
```
Comment (wking) [inline installer/pkg/config-generator/generator.go]: nit: maybe move this into the `ic` literal above?  Nothing in this line strikes me as something that couldn't happen there (e.g. no returned errors or other conditionals to deal with).

You could also move the `podCIDR`, `serviceCIDR`, and other setup to the top of this function, and then have a single:

```go
return &types.InstallConfig{
  ...
}, nil
```

at the end if they all succeed.

File: installer/pkg/config-generator/generator.go (hunk 3 above)
Comment (abhinavdahiya) [reply installer/pkg/config-generator/generator.go]: done

File: installer/pkg/config-generator/generator.go (hunk 4 above)
Comment (abhinavdahiya) [reply installer/pkg/config-generator/generator.go]: done
